File: primes-ref/primes.hs
module Main where
import System.Environment(getArgs)
import qualified Data.Map as Map
 
primes :: [Int]
primes =
  2 : (nextPrime Map.empty 3)
  where
    nextPrime :: Map.Map Int Int -> Int -> [Int]
    nextPrime sieve candidate =
      case (Map.lookup candidate sieve) of
        Nothing  -> candidate : (nextPrime (nextSieve sieve candidate) $ candidate + 2)
        Just _ -> nextPrime (nextSieve sieve candidate) $ candidate + 2
 
    nextSieve :: Map.Map Int Int -> Int -> Map.Map Int Int
    nextSieve sieve candidate =
      case (Map.lookup candidate sieve) of
        Just v -> enqueue (Map.delete candidate sieve) candidate v
        Nothing    -> enqueue sieve candidate $ candidate + candidate
 
    enqueue :: Map.Map Int Int -> Int -> Int -> Map.Map Int Int
    enqueue sieve n step =
      let m = n + step
      in case (Map.lookup m sieve) of
        Just _ -> enqueue sieve m step
        Nothing -> Map.insert m step sieve
 
main = do
  n <- getArgs
  let limit = (read (head n)) :: Int
  putStrLn $ show . last . take limit $ primes

File: primes-ref/primes.cc
#include <iostream>
#include <vector>
using namespace std;
 
vector<unsigned int> sievePrimes(unsigned int limit) {
    vector<unsigned int> primes;
    if (limit > 1) { 
        primes.push_back(2);
        vector<bool> sieve((limit + 1) / 2, false);
        for (unsigned long long i = 1, n = 3; i < sieve.size(); ++i, n += 2) {
            if (!sieve[i]) {
                primes.push_back(n);
                for (unsigned long long j = n * n / 2; j < sieve.size(); j += n) {
                    sieve[j] = true;
                }
            }
        }
    }
    return primes;
}

int main() {
    vector<unsigned int> p = sievePrimes(16000000);
    cout << p.size() << endl;
    cout << p[1000000-1] << endl;
}